File: benchmarks/tsb/bench_dataframe_rolling_agg.ts
/**
 * Benchmark: dataFrameRollingAgg on a 100k-row DataFrame
 */
import { DataFrame, Series, dataFrameRollingAgg } from "../../src/index.js";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const a = new Series(Float64Array.from({ length: ROWS }, (_, i) => Math.sin(i * 0.01)));
const b = new Series(Float64Array.from({ length: ROWS }, (_, i) => Math.cos(i * 0.01)));
const df = new DataFrame({ a, b });
const fns = {
  mean: (v: readonly number[]) => v.reduce((x, y) => x + y, 0) / v.length,
  sum: (v: readonly number[]) => v.reduce((x, y) => x + y, 0),
};

for (let i = 0; i < WARMUP; i++) {
  dataFrameRollingAgg(df, 10, fns);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  dataFrameRollingAgg(df, 10, fns);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_rolling_agg",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_dataframe_rolling_agg.py
"""Benchmark: DataFrame rolling multi-aggregation on 100k rows"""
import json, time
import numpy as np
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

df = pd.DataFrame({
    "a": np.sin(np.arange(ROWS) * 0.01),
    "b": np.cos(np.arange(ROWS) * 0.01),
})

for _ in range(WARMUP):
    df.rolling(10).agg(["mean", "sum"])

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.rolling(10).agg(["mean", "sum"])
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "dataframe_rolling_agg",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
